### PR TITLE
[viewport] Add unit tests to ViewportPlugin.

### PR DESCRIPTION
**Pull request by pengdev**
### Summary of changes

Add unit tests to the viewport plugin.

## Pull request checklist:
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog></changelog>`.
 - [ ] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt [hunk 1]
```diff
@@ -193,11 +201,13 @@ class ViewportPluginImpl : ViewportPlugin {
...
-        currentStatus,
-        reason
-      )
+      handler.post {
+        it.onViewportStatusChanged(
+          previousStatus,
+          currentStatus,
+          reason
+        )
+      }
```
Comment (pengdev) [inline plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt:210]: @macdrevx added message queue(handler on Android) to post the status updates.

File: plugin-viewport/src/test/kotlin/com/mapbox/maps/plugin/viewport/transition/TransitionUtilsTest.kt [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+import io.mockk.mockk
+import org.junit.Assert.assertEquals
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+class TransitionUtilsTest {
+  @Test
+  fun `test normalizeBearing 1`() {
```
Comment (alexander-kulikovskii) [inline plugin-viewport/src/test/kotlin/com/mapbox/maps/plugin/viewport/transition/TransitionUtilsTest.kt:15]: seems `normalize Bearing 1-8` should be parameterized tests.
other can be moved to another test class

File: plugin-viewport/src/test/kotlin/com/mapbox/maps/plugin/viewport/transition/TransitionUtilsTest.kt [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+  @Test
+  fun `test normalizeBearing 8`() {
+    val expected = -0.0
+
+    val actual = normalizeBearing(
+      currentBearing = -0.0,
+      targetBearing = 0.0
+    )
+
+    assertEquals(expected, actual, 0.000001)
```
Comment (alexander-kulikovskii) [inline plugin-viewport/src/test/kotlin/com/mapbox/maps/plugin/viewport/transition/TransitionUtilsTest.kt:107]:
```suggestion
    assertEquals(expected, actual, EPS)
```

File: plugin-viewport/src/test/kotlin/com/mapbox/maps/plugin/viewport/transition/MapboxViewportTransitionFactoryTest.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  private val mapPluginProviderDelegate = mockk<MapPluginProviderDelegate>(relaxed = true)
+  private val cameraPlugin = mockk<CameraAnimationsPlugin>()
+  private val delegateProvider = mockk<MapDelegateProvider>(relaxed = true)
+  private val animatorSet = mockk<AnimatorSet>()
+  private val constrainedSet = mockk<AnimatorSet>()
+  private lateinit var transitionsFactory: MapboxViewportTransitionFactory
+
+  @Before
+  fun setup() {
+    mockkStatic("com.mapbox.maps.plugin.viewport.transition.TransitionUtilsKt")
```
Comment (alexander-kulikovskii) [inline plugin-viewport/src/test/kotlin/com/mapbox/maps/plugin/viewport/transition/MapboxViewportTransitionFactoryTest.kt:38]: maybe move to constant?
```suggestion
    mockkStatic(TRANSITION_UTILS)
```
same for `CameraAnimationsUtils`

File: plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt [hunk 5]
```diff
@@ -37,7 +39,8 @@ import java.util.concurrent.CopyOnWriteArraySet
  *  - transitioning (camera is being managed by a ViewportTransition)
  */
 @MapboxExperimental
-class ViewportPluginImpl : ViewportPlugin {
+class ViewportPluginImpl(private val handler: Handler = Handler(Looper.getMainLooper())) :
```
Comment (alexander-kulikovskii) [inline plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt:42]: should we add `JvmOverloads`?

File: plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt [hunk 6]
```diff
@@ -105,9 +108,11 @@ class ViewportPluginImpl : ViewportPlugin {
     @Synchronized get
 
   private fun updateStatus(targetStatus: ViewportStatus, reason: String) {
-    val previousStatus = status
-    status = targetStatus
-    notifyStatusChanged(previousStatus, targetStatus, reason)
+    if (targetStatus != status) {
+      val previousStatus = status
```
Comment (alexander-kulikovskii) [inline plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt:112]: how many times this update can happen? 
if too frequent maybe move `previousStatus` outside and don't recreate it?

File: plugin-viewport/src/test/kotlin/com/mapbox/maps/plugin/viewport/transition/TransitionUtilsTest.kt (hunk 2 above)
Comment (pengdev) [reply plugin-viewport/src/test/kotlin/com/mapbox/maps/plugin/viewport/transition/TransitionUtilsTest.kt:15]: This is a direct port of unit tests from nav sdk, I will keep it in the beta release and refactor these to parameterised tests before rc release.

File: plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt (hunk 5 above)
Comment (pengdev) [reply plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt:42]: I don't think this constructor would be used as public API, as the initialisation of plugin instance is done internally in our sdk(but in another module, so need to keep this class public.)

File: plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt (hunk 6 above)
Comment (pengdev) [reply plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt:112]: it shouldn't be very frequent, will only happen if user triggers a new state transition.
Could you suggest more details how would you optimise it?

**Comment by pengdev** [comment]
@alexander-kulikovskii I will merge this PR to feature branch as the unit tests are added for all the impl classes, please continue reviewing the final PR at https://github.com/mapbox/mapbox-maps-android/pull/1033

File: plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt (hunk 6 above)
Comment (alexander-kulikovskii) [reply plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt:112]: just add `previousStatus` as variable for this class. GC wouldn't clear it as current local value

**Pull request merged**